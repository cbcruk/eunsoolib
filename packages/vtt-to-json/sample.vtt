WEBVTT
Kind: captions
Language: en

00:00:02.630 --> 00:00:08.669 align:start position:0%
 
[Music]

00:00:08.669 --> 00:00:08.679 align:start position:0%
[Music]
 

00:00:08.679 --> 00:00:12.150 align:start position:0%
[Music]
Yeah,<00:00:09.679><c> I</c><00:00:09.920><c> love</c><00:00:10.080><c> you.</c><00:00:11.040><c> I'm</c><00:00:11.280><c> in</c><00:00:11.440><c> a</c><00:00:11.599><c> funk,</c><00:00:11.840><c> so</c><00:00:12.000><c> I</c>

00:00:12.150 --> 00:00:12.160 align:start position:0%
Yeah, I love you. I'm in a funk, so I
 

00:00:12.160 --> 00:00:13.830 align:start position:0%
Yeah, I love you. I'm in a funk, so I
bought<00:00:12.320><c> a</c><00:00:12.480><c> bouquet</c><00:00:12.800><c> of</c><00:00:12.960><c> roses</c><00:00:13.280><c> and</c><00:00:13.440><c> cut</c><00:00:13.679><c> them</c>

00:00:13.830 --> 00:00:13.840 align:start position:0%
bought a bouquet of roses and cut them
 

00:00:13.840 --> 00:00:15.030 align:start position:0%
bought a bouquet of roses and cut them
up<00:00:14.000><c> at</c><00:00:14.160><c> your</c><00:00:14.320><c> doorstep.</c><00:00:14.719><c> Your</c><00:00:14.880><c> new</c>

00:00:15.030 --> 00:00:15.040 align:start position:0%
up at your doorstep. Your new
 

00:00:15.040 --> 00:00:16.950 align:start position:0%
up at your doorstep. Your new
neighborhood<00:00:15.440><c> is</c><00:00:15.679><c> gorgeous.</c><00:00:16.160><c> I</c><00:00:16.400><c> paid</c><00:00:16.640><c> a</c><00:00:16.800><c> lot</c>

00:00:16.950 --> 00:00:16.960 align:start position:0%
neighborhood is gorgeous. I paid a lot
 

00:00:16.960 --> 00:00:19.029 align:start position:0%
neighborhood is gorgeous. I paid a lot
of<00:00:17.039><c> money</c><00:00:17.279><c> for</c><00:00:17.520><c> the</c><00:00:17.680><c> fragrances</c><00:00:18.160><c> you</c><00:00:18.880><c> were</c>

00:00:19.029 --> 00:00:19.039 align:start position:0%
of money for the fragrances you were
 

00:00:19.039 --> 00:00:20.870 align:start position:0%
of money for the fragrances you were
dating,<00:00:19.359><c> and</c><00:00:19.520><c> I</c><00:00:19.680><c> sold</c><00:00:19.920><c> some</c><00:00:20.080><c> lemonade</c><00:00:20.560><c> just</c><00:00:20.720><c> to</c>

00:00:20.870 --> 00:00:20.880 align:start position:0%
dating, and I sold some lemonade just to
 

00:00:20.880 --> 00:00:22.710 align:start position:0%
dating, and I sold some lemonade just to
afford<00:00:21.199><c> them.</c><00:00:21.600><c> I</c><00:00:21.760><c> know</c><00:00:21.920><c> it's</c><00:00:22.160><c> not</c><00:00:22.320><c> a</c><00:00:22.480><c> really</c>

00:00:22.710 --> 00:00:22.720 align:start position:0%
afford them. I know it's not a really
 

00:00:22.720 --> 00:00:24.790 align:start position:0%
afford them. I know it's not a really
good<00:00:22.960><c> occasion</c><00:00:23.359><c> to</c><00:00:23.519><c> be</c><00:00:23.760><c> barging</c><00:00:24.160><c> in.</c><00:00:24.560><c> I</c>

00:00:24.790 --> 00:00:24.800 align:start position:0%
good occasion to be barging in. I
 

00:00:24.800 --> 00:00:26.310 align:start position:0%
good occasion to be barging in. I
couldn't<00:00:25.039><c> help</c><00:00:25.199><c> but</c><00:00:25.439><c> watch</c><00:00:25.600><c> you</c><00:00:25.760><c> kiss</c><00:00:26.000><c> her</c><00:00:26.160><c> by</c>

00:00:26.310 --> 00:00:26.320 align:start position:0%
couldn't help but watch you kiss her by
 

00:00:26.320 --> 00:00:28.390 align:start position:0%
couldn't help but watch you kiss her by
the<00:00:26.480><c> kitchen</c><00:00:26.800><c> sink.</c><00:00:27.199><c> I</c><00:00:27.439><c> swung</c><00:00:27.680><c> the</c><00:00:27.840><c> door</c><00:00:28.080><c> open,</c>

00:00:28.390 --> 00:00:28.400 align:start position:0%
the kitchen sink. I swung the door open,
 

00:00:28.400 --> 00:00:30.470 align:start position:0%
the kitchen sink. I swung the door open,
tippy<00:00:28.800><c> toeed,</c><00:00:29.119><c> farther</c><00:00:29.519><c> in.</c><00:00:30.000><c> I</c><00:00:30.240><c> wasn't</c>

00:00:30.470 --> 00:00:30.480 align:start position:0%
tippy toeed, farther in. I wasn't
 

00:00:30.480 --> 00:00:32.150 align:start position:0%
tippy toeed, farther in. I wasn't
crying,<00:00:30.880><c> I</c><00:00:31.039><c> was</c><00:00:31.199><c> staring</c><00:00:31.439><c> and</c><00:00:31.679><c> forgot</c><00:00:32.000><c> to</c>

00:00:32.150 --> 00:00:32.160 align:start position:0%
crying, I was staring and forgot to
 

00:00:32.160 --> 00:00:34.310 align:start position:0%
crying, I was staring and forgot to
blink.<00:00:32.640><c> She</c><00:00:32.800><c> saw</c><00:00:33.040><c> me</c><00:00:33.200><c> standing</c><00:00:33.520><c> by</c><00:00:33.680><c> the</c><00:00:33.920><c> TV</c><00:00:34.160><c> and</c>

00:00:34.310 --> 00:00:34.320 align:start position:0%
blink. She saw me standing by the TV and
 

00:00:34.320 --> 00:00:36.069 align:start position:0%
blink. She saw me standing by the TV and
she<00:00:34.559><c> wouldn't</c><00:00:34.880><c> stop</c><00:00:35.120><c> screaming.</c><00:00:35.600><c> So</c><00:00:35.760><c> I</c><00:00:35.920><c> tried</c>

00:00:36.069 --> 00:00:36.079 align:start position:0%
she wouldn't stop screaming. So I tried
 

00:00:36.079 --> 00:00:37.590 align:start position:0%
she wouldn't stop screaming. So I tried
to<00:00:36.239><c> be</c><00:00:36.320><c> discreet</c><00:00:36.719><c> and</c><00:00:36.880><c> told</c><00:00:37.040><c> her,</c><00:00:37.120><c> "Calm</c><00:00:37.440><c> your</c>

00:00:37.590 --> 00:00:37.600 align:start position:0%
to be discreet and told her, "Calm your
 

00:00:37.600 --> 00:00:39.590 align:start position:0%
to be discreet and told her, "Calm your
tits."<00:00:38.079><c> She</c><00:00:38.239><c> grabbed</c><00:00:38.480><c> the</c><00:00:38.640><c> kitchen</c><00:00:38.960><c> knife.</c><00:00:39.440><c> So</c>

00:00:39.590 --> 00:00:39.600 align:start position:0%
tits." She grabbed the kitchen knife. So
 

00:00:39.600 --> 00:00:41.430 align:start position:0%
tits." She grabbed the kitchen knife. So
I<00:00:39.760><c> pulled</c><00:00:39.920><c> out</c><00:00:40.079><c> the</c><00:00:40.239><c> blick</c><00:00:40.640><c> and</c><00:00:40.879><c> got</c><00:00:41.120><c> it</c><00:00:41.280><c> all</c>

00:00:41.430 --> 00:00:41.440 align:start position:0%
I pulled out the blick and got it all
 

00:00:41.440 --> 00:00:43.590 align:start position:0%
I pulled out the blick and got it all
the<00:00:41.680><c> time.</c><00:00:42.000><c> Thank</c><00:00:42.239><c> God</c><00:00:42.480><c> I</c><00:00:42.640><c> did</c><00:00:42.800><c> for</c><00:00:43.040><c> this</c><00:00:43.360><c> cuz</c>

00:00:43.590 --> 00:00:43.600 align:start position:0%
the time. Thank God I did for this cuz
 

00:00:43.600 --> 00:00:45.670 align:start position:0%
the time. Thank God I did for this cuz
she<00:00:43.840><c> was</c><00:00:44.000><c> seeing</c><00:00:44.320><c> bread</c><00:00:44.719><c> and</c><00:00:44.960><c> all</c><00:00:45.120><c> I</c><00:00:45.280><c> saw</c><00:00:45.520><c> was</c>

00:00:45.670 --> 00:00:45.680 align:start position:0%
she was seeing bread and all I saw was
 

00:00:45.680 --> 00:00:47.990 align:start position:0%
she was seeing bread and all I saw was
you.<00:00:46.079><c> It</c><00:00:46.399><c> happened</c><00:00:46.640><c> in</c><00:00:46.879><c> the</c><00:00:47.039><c> flash.</c><00:00:47.600><c> When</c><00:00:47.840><c> she</c>

00:00:47.990 --> 00:00:48.000 align:start position:0%
you. It happened in the flash. When she
 

00:00:48.000 --> 00:00:50.069 align:start position:0%
you. It happened in the flash. When she
charged<00:00:48.399><c> at</c><00:00:48.559><c> me,</c><00:00:48.719><c> y'all</c><00:00:49.039><c> crisscross,</c><00:00:49.680><c> saw</c><00:00:49.920><c> her</c>

00:00:50.069 --> 00:00:50.079 align:start position:0%
charged at me, y'all crisscross, saw her
 

00:00:50.079 --> 00:00:51.910 align:start position:0%
charged at me, y'all crisscross, saw her
fall<00:00:50.399><c> to</c><00:00:50.559><c> the</c><00:00:50.719><c> floor.</c><00:00:51.039><c> Then</c><00:00:51.200><c> you</c><00:00:51.360><c> paused</c><00:00:51.760><c> and</c>

00:00:51.910 --> 00:00:51.920 align:start position:0%
fall to the floor. Then you paused and
 

00:00:51.920 --> 00:00:53.950 align:start position:0%
fall to the floor. Then you paused and
in<00:00:52.160><c> horror</c><00:00:52.559><c> that</c><00:00:52.800><c> shot</c><00:00:53.039><c> wasn't</c>

00:00:53.950 --> 00:00:53.960 align:start position:0%
in horror that shot wasn't
 

00:00:53.960 --> 00:00:58.360 align:start position:0%
in horror that shot wasn't
for<00:00:54.960><c> it.</c>

00:00:58.360 --> 00:00:58.370 align:start position:0%
 
 

00:00:58.370 --> 00:01:01.460 align:start position:0%
 
[Music]

