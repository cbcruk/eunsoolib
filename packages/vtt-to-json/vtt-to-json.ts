export type Cue = {
  seconds: number
  timestamp: string
  text: string
}

export class VttParser {
  private data: string

  private state = {
    timeStamp: '',
    lastText: '',
  }

  constructor(data: string) {
    this.data = data
  }

  private getSeconds(timestamp: string): number {
    const parts = timestamp.split(':')

    if (parts.length !== 3) {
      return 0
    }

    const [h, m, s] = parts

    return parseInt(h, 10) * 3600 + parseInt(m, 10) * 60 + parseFloat(s)
  }

  toJson(): Cue[] {
    const cues: Cue[] = []
    const lines = this.data
      .split('\n')
      .map((line) => line.trim())
      .filter(Boolean)

    for (const line of lines) {
      // 타임스탬프 줄 확인: "00:00:01.000 --> 00:00:05.000" 형식
      const match = line.match(/^(\d{2}:\d{2}:\d{2}\.\d{3}) --> /)

      if (match) {
        // 타임스탬프 줄이면 현재 자막 시작 시간으로 설정하고, 다음 줄부터 자막으로 인식
        this.state.timeStamp = match[1]
        continue
      }

      if (
        // 유효한 타임스탬프가 없는 경우
        !this.state.timeStamp ||
        // 이전 자막과 동일한 텍스트인 경우
        line === this.state.lastText ||
        // 스타일 태그인 경우
        /<\/c>$/.test(line)
      ) {
        continue
      }

      cues.push({
        seconds: this.getSeconds(this.state.timeStamp),
        timestamp: this.state.timeStamp,
        text: line,
      })

      // 마지막 자막 텍스트 저장 (중복 제거)
      this.state.lastText = line
    }

    return cues
  }
}
