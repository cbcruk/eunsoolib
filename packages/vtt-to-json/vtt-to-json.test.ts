import { expect, test } from 'vitest'
import { VttParser } from './vtt-to-json'
import { readFile } from 'fs/promises'

test('vttToJson', async () => {
  const vtt = await readFile('packages/vtt-to-json/sample.vtt', 'utf-8')
  const parser = new VttParser(vtt)
  const result = parser.toJson()

  expect(result).toMatchSnapshot()
})
