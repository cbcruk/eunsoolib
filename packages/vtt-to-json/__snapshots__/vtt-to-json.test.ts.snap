// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`vttToJson 1`] = `
[
  {
    "seconds": 2.63,
    "text": "[Music]",
    "timestamp": "00:00:02.630",
  },
  {
    "seconds": 12.15,
    "text": "Yeah, I love you. I'm in a funk, so I",
    "timestamp": "00:00:12.150",
  },
  {
    "seconds": 13.83,
    "text": "bought a bouquet of roses and cut them",
    "timestamp": "00:00:13.830",
  },
  {
    "seconds": 15.03,
    "text": "up at your doorstep. Your new",
    "timestamp": "00:00:15.030",
  },
  {
    "seconds": 16.95,
    "text": "neighborhood is gorgeous. I paid a lot",
    "timestamp": "00:00:16.950",
  },
  {
    "seconds": 19.029,
    "text": "of money for the fragrances you were",
    "timestamp": "00:00:19.029",
  },
  {
    "seconds": 20.87,
    "text": "dating, and I sold some lemonade just to",
    "timestamp": "00:00:20.870",
  },
  {
    "seconds": 22.71,
    "text": "afford them. I know it's not a really",
    "timestamp": "00:00:22.710",
  },
  {
    "seconds": 24.79,
    "text": "good occasion to be barging in. I",
    "timestamp": "00:00:24.790",
  },
  {
    "seconds": 26.31,
    "text": "couldn't help but watch you kiss her by",
    "timestamp": "00:00:26.310",
  },
  {
    "seconds": 28.39,
    "text": "the kitchen sink. I swung the door open,",
    "timestamp": "00:00:28.390",
  },
  {
    "seconds": 30.47,
    "text": "tippy toeed, farther in. I wasn't",
    "timestamp": "00:00:30.470",
  },
  {
    "seconds": 32.15,
    "text": "crying, I was staring and forgot to",
    "timestamp": "00:00:32.150",
  },
  {
    "seconds": 34.31,
    "text": "blink. She saw me standing by the TV and",
    "timestamp": "00:00:34.310",
  },
  {
    "seconds": 36.069,
    "text": "she wouldn't stop screaming. So I tried",
    "timestamp": "00:00:36.069",
  },
  {
    "seconds": 37.59,
    "text": "to be discreet and told her, "Calm your",
    "timestamp": "00:00:37.590",
  },
  {
    "seconds": 39.59,
    "text": "tits." She grabbed the kitchen knife. So",
    "timestamp": "00:00:39.590",
  },
  {
    "seconds": 41.43,
    "text": "I pulled out the blick and got it all",
    "timestamp": "00:00:41.430",
  },
  {
    "seconds": 43.59,
    "text": "the time. Thank God I did for this cuz",
    "timestamp": "00:00:43.590",
  },
  {
    "seconds": 45.67,
    "text": "she was seeing bread and all I saw was",
    "timestamp": "00:00:45.670",
  },
  {
    "seconds": 47.99,
    "text": "you. It happened in the flash. When she",
    "timestamp": "00:00:47.990",
  },
  {
    "seconds": 50.069,
    "text": "charged at me, y'all crisscross, saw her",
    "timestamp": "00:00:50.069",
  },
  {
    "seconds": 51.91,
    "text": "fall to the floor. Then you paused and",
    "timestamp": "00:00:51.910",
  },
  {
    "seconds": 53.95,
    "text": "in horror that shot wasn't",
    "timestamp": "00:00:53.950",
  },
  {
    "seconds": 58.37,
    "text": "[Music]",
    "timestamp": "00:00:58.370",
  },
]
`;
